[#37009] Search-based selection of participants in meetings for projects with very many project members

mrmir opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->
https://community.openproject.org/work_packages/37009

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

- Update the participants dialog to use styling similar to the share dialogs
- Keep the participant "invited" flag for now
- Add a new controller to handle participant specific actions, since there's more of them now

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

mrmir committed:
Rework participants components

mrmir committed:
Add and update specs

mrmir committed:
Update icon for ShareRowComponent

klaustopher commented:
Generic question. Why is every component in the `side_panels` subfolder? The dialog is triggered from the side panel but it's not really a component for the sidebar.

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/attended_toggle_button_component.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    def label
+      @participant.attended? ? "Attended" : "Mark as attended"
```
Those should be put into translations

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.html.erb:
```diff
@@ -0,0 +1,104 @@
+                    )
+                  ) do |button|
+                    button.with_leading_visual_icon(icon: :"op-person-assigned")
+                    "Mark all as attended"
```
Extract translation

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.html.erb:
```diff
@@ -0,0 +1,104 @@
+            box.with_row do
+              render(Primer::Beta::Blankslate.new) do |blankslate|
+                blankslate.with_visual_icon(icon: :people, size: :medium)
+                blankslate.with_heading(tag: :h2).with_content("No participants")
```
extract translation

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.html.erb:
```diff
@@ -0,0 +1,104 @@
+                blankslate.with_description do
+                  flex_layout do |flex|
+                    flex.with_row(mb: 2) do
+                      render(Primer::Beta::Text.new(color: :subtle)) { "There are no participants yet." }
```
extract translation

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.html.erb:
```diff
@@ -0,0 +1,104 @@
+                                tag: :a,
+                                href: meeting_participant_path(@meeting, participant),
+                                icon: :x,
+                                aria: { label: "Remove participant" },
```
extract translation

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.html.erb:
```diff
@@ -0,0 +1,104 @@
+        end
+      end
+    end
+  end
```
This component feels super long and doing many things at once. It feels like the row should at least be it's own component. There is quite some logic nested in here and I think it would make sense to move that into it's own thing

mrmir replied:
Yep, good call

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/list_component.rb:
```diff
@@ -0,0 +1,50 @@
+
+    def show_mark_all_attended?
+      @meeting.participants.where(attended: true).count != @meeting.participants.count && @meeting.in_progress?
+    end
```
We will need to load all participants anyways, wo we could avoid issuing two count queries here:

```ruby
def meeting_participants
  @meeting_participants ||= @meeting.participants
end

def show_mark_all_attended?
  meeting_participants.all?(&:attended?) && @meeting.in_progress?
end
```

Then we could also use the `meeting_participants` methods in the loop in the view as well instead of reloading it using `@meeting.participants`

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/manage_participants_dialog.html.erb:
```diff
@@ -0,0 +1,52 @@
+      Primer::Alpha::Dialog.new(
+        id: "manage-participants-dialog",
+        size: :large,
+        title: "Manage participants",
```
extract translations

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/manage_participants_dialog.html.erb:
```diff
@@ -0,0 +1,52 @@
+          end
+
+          container.with_row(mb: 3) do
+            render(Primer::Beta::Text.new) { "Search for and add project members as participants to this meeting." }
```
extract translations

klaustopher commented on modules/meeting/app/components/meetings/side_panel/participants/manage_participants_dialog.html.erb:
```diff
@@ -0,0 +1,52 @@
+              end
+            end
+          end
+        end
```
Why is the button only shown when the meeting is editable? You are not undoing any changes here. So IMHO the button should just be a close button and be shown every time.

klaustopher commented on modules/meeting/app/forms/meeting/participant.rb:
```diff
@@ -0,0 +1,73 @@
+
+  def filters
+    list = [
+      { name: "type", operator: "=", values: %w[User Group] },
```
Can we add a group as a participant? Is that handled correctly?

mrmir replied:
Good catch, I tested it and it doesn't really work with groups for notifications and things as well, so I've removed that option. I also removed it for the presenter field for agenda items, which is what I copied over initially, but it breaks there too

klaustopher reviewed (changes requested):
Great work so far. There are currently no translations being used and all the strings are hard coded. Those should be moved to the translation files.

Also it feels like the `list_component` does too much in general. It would help if we could extract at least the rows into their own component.

Clicking around through the dialog, I also noticed a bunch of things

- <img width="657" height="452" alt="Image" src="https://github.com/user-attachments/assets/acf6a87d-91c4-4cc9-b6ce-cc26cbe15012" />
  The presenter label here is not correct. Also, do we really want scrolling in the view even with the empty state? I think we talked about the dialog being longer?
- Mentioned this inline, but the cancel button feels really weird when it is not cancelling anything. We should label it with Close. In the mockup that is linked in the ticket there is not even a footer. Maybe we should recheck with Parimal for this.

Rest looks fine. I would also consider moving the dialog components out of the `Meetings::SidePanel` namespace as the dialog itself has no direct connection to the sidebar in my opinion. They are triggered from the sidebar, but I think the dialog can just be `Meetings::Participants` but that's just a nitpick, feel free to ignore :)

mrmir committed:
Incorporate PR feedback

mrmir commented:
Thanks for the feedback! 

Can't believe I forgot to add translations, but those should all be fixed now hopefully, and made consistent for all older participants related strings as well.

I moved everything out of the `SidePanel` namespace and I've also broken up the `list_component` and separated out the header and row components, which cleans it up a little I think.

As for the dialog, I've fixed the placeholder text and went back and forth with Parimal to come up with an implementation that works:

- The `xlarge` variant is used to prevent scrolling with the initial/empty states
- A min-height is set on the body to always have the same height for the modal
- The footer button is removed

Could you please take another look?

klaustopher commented on modules/meeting/app/components/meetings/participants/list_component.html.erb:
```diff
@@ -0,0 +1,44 @@
+            end
+          else
+            render(Primer::BaseComponent.new(tag: :div, classes: "participants-list")) do
+              @meeting.all_changeable_participants.sort.each do |user|
+                if @meeting.participants.present? && participant = @meeting.participants.detect { |p| p.user_id == user.id }
```
I still find this construct a bit odd. We cannot add a participant that is not invited or attended, so what is the difference here between just iterating over `@meeting.participants`?

mrmir replied:
Thanks for pointing this out, it was from the old implementation and I overlooked it. Removed now, along with some other unnecessary participants grouping methods 👍🏽

klaustopher reviewed (approved):
Changes look good.

I left one more comment inline. I think we can get rid of the `all_changeable_participants` thing now. We cannot create a meeting participation that is not at least `invited`. When a person is removed from the meeting we delete the participant record. So there cannot be a `Participant` that is not `invited=true`. And thus we can just replace it with `@meeting.participants`
